pkg/frontend/querymiddleware/split_by_interval_test.go

```diff
@@ -0,0 +1,161 @@
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sort"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/mimir/pkg/util"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/user"
+)
+
+func TestQuerySplittingCorrectness(t *testing.T) {
+	var (
+		numSeries          = 1000
+		numStaleSeries     = 100
+		numHistograms      = 1000
+		numStaleHistograms = 100
+		histogramBuckets   = []float64{1.0, 2.0, 4.0, 10.0, 100.0, math.Inf(1)}
+	)
+
+	tests := map[string]struct {
```

Comment:
I would like to see more complex queries here as well. You can get some inspiration from real queries you found during your analysis.